dvc/objects/db/reference.py

```diff
@@ -60,25 +58,19 @@ def _add_file(
         hash_info: "HashInfo",
         move: bool = False,
     ):
+        from dvc import fs
+
         self.makedirs(to_info.parent)
         if hash_info.isdir:
             return super()._add_file(
-                from_fs, from_info, to_info, hash_info, move
+                from_fs, from_info, to_info, hash_info, move=move
             )
+
         ref_file = ReferenceHashFile(from_info, from_fs, hash_info)
         self._obj_cache[hash_info] = ref_file
-        ref_fobj = io.BytesIO(ref_file.to_bytes())
-        ref_fobj.seek(0)
-        try:
-            self.fs.upload_fobj(ref_fobj, to_info)
-        except OSError as exc:
-            if isinstance(exc, FileExistsError) or (
-                os.name == "nt"
-                and exc.__context__
-                and isinstance(exc.__context__, FileExistsError)
-            ):
-                logger.debug("'%s' file already exists, skipping", to_info)
-            else:
-                raise
+        content = ref_file.to_bytes()
+        fs.utils.transfer(
+            from_fs, from_info, self.fs, to_info, move=move, content=content
+        )
```

Comment:
@pmrowla, what do you think of this? Is `transfer` needed here, or should it use more primitive functions like `pipe_file`?

Reply:
The `upload_fobj` for reference objects isn't really a file transfer, it's just a `write()` of the serialized reference data from memory.

It seems to me like we should still have a distinction between a file transfer and a direct binary data write/upload, since in this case there really isn't a `from_fs` and `from_info` that we are transferring.

One alternative would maybe be adding something like `ReferenceHashFile.to_memfile()` where it returns a `memfs://some_tmp_pathname` path to a `MemoryFileSystem` entry (containing the serialized data). And then we could do an actual "file transfer" using the memfs and memfs_path_info without needing the `content` field?

Reply:
yeah, initially I started with two methods, but at the end `fs.utils.write` is going to be equivalent to `upload_fobj`, so I don't see much difference there. With `transfer`, I tried to push the problem of getting `size` for the callback support inside the `transfer` itself, rather than putting that burden on the caller.

We do the following when transferring files straight to the remote, which is just a file transfer at the end:
```python
 with fs.open(path_info, mode="rb", chunk_size=fs.CHUNK_SIZE) as stream:
         stream = HashedStreamReader(stream)
         upload_odb.fs.upload_fobj(
             stream, tmp_info, desc=path_info.name, size=fs.getsize(path_info)
         )
```

But maybe the best thing to do here is put that burden on the caller, rather than trying to generalize it in `transfer`.